drivers/i2s/i2s_litex.c

```diff
@@ -0,0 +1,617 @@
+/*
+ * Copyright (c) 2020 Antmicro <www.antmicro.com>
+ * 
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <drivers/i2s.h>
+#include <sys/byteorder.h>
+#include <soc.h>
+#include <sys/util.h>
+#include <assert.h>
+#include "i2s_litex.h"
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(i2s_litex);
+#define DEV_CFG(dev) ((struct i2s_litex_cfg *const)(dev)->config_info)
+#define DEV_DATA(dev) ((struct i2s_litex_data *const)(dev)->driver_data)
+
+#define MODULO_INC(val, max)                                                   \
+	{                                                                      \
+		val = (++val < max) ? val : 0;                                 \
+	}
+
+/**
+ * @brief Enable i2s device
+ *
+ * @param reg base register of device 
+ *
+ * @return N/A
+ */
+static void i2s_enable(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_CONTROL_REG_OFFSET);
+
+	litex_write8(reg_data | I2S_ENABLE, reg + I2S_CONTROL_REG_OFFSET);
+}
+
+/**
+ * @brief Disable i2s device
+ *
+ * @param reg base register of device
+ *
+ * @return N/A
+ */
+static void i2s_disable(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_CONTROL_REG_OFFSET);
+
+	litex_write8(reg_data & ~(I2S_ENABLE), reg + I2S_CONTROL_REG_OFFSET);
+}
+
+/**
+ * @brief Reset i2s fifo 
+ *
+ * @param reg base register of device 
+ *
+ * @return N/A
+ */
+static void i2s_reset_fifo(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_CONTROL_REG_OFFSET);
+
+	litex_write8(reg_data | I2S_FIFO_RESET, reg + I2S_CONTROL_REG_OFFSET);
+}
+
+/**
+ * @brief Get i2s format handled by device 
+ *
+ * @param reg base register of device 
+ *
+ * @return currently supported format
+ */
+static int i2s_get_foramt(int reg)
```

Comment:
optional: returning an 'enum i2s_fmt' (if there is such a thing) here makes the types easier to debug.

Reply:
I want to be compatible with i2s standard header and reuse defined variables.
Unfortunately, they are defined as #define. The type for these values comes with i2s header.
I will change returned type to i2s_fmt_t.